pkg/3scale/amp/operator/backend_options_provider.go

```diff
@@ -0,0 +1,144 @@
+package operator
+
+import (
+	"fmt"
+
+	"github.com/3scale/3scale-operator/pkg/3scale/amp/component"
+	appsv1alpha1 "github.com/3scale/3scale-operator/pkg/apis/apps/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type OperatorBackendOptionsProvider struct {
+	apimanagerSpec *appsv1alpha1.APIManagerSpec
+	namespace      string
+	client         client.Client
+	backendOptions *component.BackendOptions
+	secretSource   *SecretSource
+}
+
+func NewOperatorBackendOptionsProvider(apimanagerSpec *appsv1alpha1.APIManagerSpec, namespace string, client client.Client) *OperatorBackendOptionsProvider {
+	return &OperatorBackendOptionsProvider{
+		apimanagerSpec: apimanagerSpec,
+		namespace:      namespace,
+		client:         client,
+		backendOptions: component.NewBackendOptions(),
+		secretSource:   NewSecretSource(client, namespace),
+	}
+}
+
+func (o *OperatorBackendOptionsProvider) GetBackendOptions() (*component.BackendOptions, error) {
+	o.backendOptions.AppLabel = *o.apimanagerSpec.AppLabel
+	o.backendOptions.TenantName = *o.apimanagerSpec.TenantName
+	o.backendOptions.WildcardDomain = o.apimanagerSpec.WildcardDomain
+
+	err := o.setSecretBasedOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	o.setResourceRequirementsOptions()
+	o.setReplicas()
+
+	err = o.backendOptions.Validate()
+	return o.backendOptions, err
+}
+
+func (o *OperatorBackendOptionsProvider) setSecretBasedOptions() error {
+	cases := []struct {
+		field       *string
+		secretName  string
+		secretField string
+		defValue    string
```

Comment:
What if the field has no default value because it is required? setting it as "" might not be a solution because there are some fields where the default value is really a "" and would not mean, 'not having a default' (distinguish vs default value and value not provided)

Reply:
This is implementation for the `backend_options_provider`. Does not apply. As explained earlier,  other option providers may require existence of a fields, so default values would not make sense at all and other public methods of `SecretSource` will be needed, not having any argument for the default value.